src/peppr/rotamer.py

```diff
@@ -0,0 +1,1439 @@
+
+    References
+    ----------
+    .. [1] https://pmc.ncbi.nlm.nih.gov/articles/PMC4983197/
+    """
+
+    rama_scores: list[ResidueRamaScore] = field(
+        default_factory=list,
+        metadata={
+            "description": "List of ramachandran scores for each residue in the structure."
+        },
+    )
+
+    @classmethod
+    def from_atoms(cls, atom_array: struc.AtomArray) -> "RamaScore":
+        """
+        Create RamaScore from a protein structure.
+
+        Parameters
+        ----------
+        atom_array : struc.AtomArray
+            The structure.
+
+        Returns
+        -------
+        RamaScore
+            An instance of RamaScore containing the ramachandran scores for the structure.
+        """
+        if isinstance(atom_array, struc.AtomArray):
+            rama_scores = _classify_phi_psi(atom_array)
+            return cls(rama_scores=rama_scores)
+        else:
+            raise TypeError("atom_array must be an AtomArray")
+
+
+def get_fraction_of_rotamer_outliers(atom_array: struc.AtomArray) -> float:
+    """
+    Compute the fraction of rotamer outliers for given structure.
+
+    Parameters
+    ----------
+    atom_array : struc.AtomArray
+        The structure.
+
+    Returns
+    -------
+    float
+        The fraction of rotamer outliers in the structure, calculated as the number of outlier
+        rotamers divided by the total number of rotamers.
+    """
+    outlier_rotamers = 0
+    total_rotamers = 0
+    result = RotamerScore.from_atoms(atom_array)
+    for rotamer_score in result.rotamer_scores:
+        if isinstance(rotamer_score, list):
+            # If it's a stack, we need to iterate through each score
+            for score in rotamer_score:
+                if score.classification == ConformerClass.OUTLIER:
+                    outlier_rotamers += 1
+                total_rotamers += 1
+        elif isinstance(rotamer_score, ResidueRotamerScore):
+            # If it's a single score, check its classification
+            if rotamer_score.classification == ConformerClass.OUTLIER:
+                outlier_rotamers += 1
+            total_rotamers += 1
+    return outlier_rotamers / total_rotamers if total_rotamers > 0 else 0.0
+
+
+def get_fraction_of_rama_outliers(
+    atom_array: struc.AtomArray | struc.AtomArrayStack,
+) -> float:
+    """
+    Compute the fraction of ramachandran outliers for given structure.
+
+    Parameters
+    ----------
+    atom_array : struc.AtomArray | struc.AtomArrayStack
+        The AtomArray or AtomArrayStack containing the structure(s).
+
+    Returns
+    -------
+    float
+        The fraction of ramachandran outliers in the structure, calculated as the number of outlier
+        ramachandran angles divided by the total number of ramachandran angles.
+    """
+    outlier_rama = 0
+    total_rama = 0
+    result = RamaScore.from_atoms(atom_array)
+
+    for rama_score in result.rama_scores:
+        if isinstance(rama_score, list):
```

Comment:
This branch can now be removed, now that `RamaScore.from_atoms()` only processes single `AtomArray`s